src/transformers/models/llama/modeling_llama.py

```diff
@@ -352,18 +354,24 @@ def forward(
         key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
+        if isinstance(past_key_value, Cache):
+            past_key_value.update_pre_rotation(key_states, value_states)
+
         kv_seq_len = key_states.shape[-2]
-        if past_key_value is not None:
+        if past_key_value:
             kv_seq_len += past_key_value[0].shape[-2]
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
-        if past_key_value is not None:
+        if past_key_value:
             # reuse k, v, self_attention
             key_states = torch.cat([past_key_value[0], key_states], dim=2)
             value_states = torch.cat([past_key_value[1], value_states], dim=2)
 
-        past_key_value = (key_states, value_states) if use_cache else None
+        if isinstance(past_key_value, Cache):
+            past_key_value.update(key_states, value_states)
+        elif past_key_value is not None:
+            past_key_value = (key_states, value_states)
```

Comment:
```suggestion
       past_key_value.update(key_states, value_states)
```

I was envisioning to just assume that the `past_key_value` is **always** of type `Cache` at this point

Reply:
This shoudn't be backwards breaking no?